// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This file defines the overall structure of a P4 program in our IR, as well
// as the supported statements and expressions.
// The structure is tailored to be easy to consume rather than produce.
// It deviates from the underlying bmv2 JSON format, which is mimicked closely
// by p4_symbolic/bmv2/bmv2.proto.
// The bmv2 proto is used to initially parse the JSON format, which is then
// processed by our tool and transformed into the IR.
// The IR does not cover all the features of P4 programs.
// Specifically, we intentionally ignore de-parsers, stacks, unions, checksum
// calculations, and extern definitions.
// Additionally, some types of expressions and statements are unsupported, these
// are described in the relevant sections of this file.
// See the bmv2 JSON format reference for more information:
// https://github.com/p4lang/behavioral-model/blob/main/docs/JSON_format.md

syntax = "proto3";

package p4_symbolic.ir;

import "p4_pdpi/ir.proto";
import "p4_symbolic/bmv2/bmv2.proto";

// Overall structure of a p4 program.
message P4Program {
  // Header definitions, keyed by a header instance name, which is mapped
  // to its HeaderType.
  // e.g. "standard_metadata" => <standard_metadata_t type definition>.
  map<string, HeaderType> headers = 1;
  // Action definitions and implementations, keyed by qualified action name.
  map<string, Action> actions = 2;
  // Table definitions and implementations, keyed by qualified table name.
  map<string, Table> tables = 3;
  // Conditionals, keyed by their name.
  map<string, Conditional> conditionals = 4;
  // Pipelines, keyed by their name.
  map<string, Pipeline> pipeline = 5;
  // Parsers, keyed by parser names.
  map<string, Parser> parsers = 6;
  // Errors, keyed by error names.
  map<string, Error> errors = 7;

  // TODO: If needed in the future, action calls can be added here, for
  // action calls that are not wrapped in other control constructs.
  // https://github.com/p4lang/behavioral-model/blob/main/docs/JSON_format.md#pipelines
}

// A header type definition.
message HeaderType {
  string name = 1;
  int32 id = 2;
  // Maps a field name to its definition.
  map<string, HeaderField> fields = 3;
}

// A P4 header type definition.
message HeaderField {
  string name = 1;
  int32 bitwidth = 2;
  bool signed = 3;
}

// A P4 parser definition.
message Parser {
  // The name of the parser.
  string name = 1;
  // The name of the initial state of the state machine, usually "start".
  string initial_state = 2;
  // Defines all parse states of the state machine.
  map<string, ParseState> parse_states = 3;
}

// Defines a parse state.
// A parse state defines a state and its outgoing transitions in the parser's
// state machine.
message ParseState {
  string name = 1;
  // Operations performed in the parse state in the order defined by the P4
  // program.
  repeated ParserOperation parser_ops = 2;
  // Fields of the transition key for the parser state (usually just one field).
  repeated ParserTransitionKeyField transition_key_fields = 3;
  // All transitions from this state to the next states.
  repeated ParserTransition transitions = 4;
  // Control flow information for optimized symbolic execution.
  OptimizedSymbolicExecutionInfo optimized_symbolic_execution_info = 5;
}

// Defines a parser operation, which corresponds to a single statement in a
// parse state. For instance, a parser operation can be a "set" operation
// (corresponding to an assignment statement), an "extract" operation
// (corresponding to an "extract" method call), or other primitives supported by
// the target implementation.
//
// Reference:
// https://github.com/p4lang/behavioral-model/blob/main/docs/JSON_format.md#parser-operations.
//
// Note that only "extract" and "set" operations are supported for now since
// others are not required at the moment. To support more complicated parsers,
// other operations may be supported in the future.
message ParserOperation {
  // Defines an "extract" parser operation, which performs extraction for a
  // fixed-width header.
  message Extract {
    // The name of the header instance to be extracted.
    HeaderValue header = 1;
  }

  // Defines a "set" parser operation, which sets a field with the given value.
  message Set {
    // The field whose value is to be updated with the rvalue.
    FieldValue lvalue = 1;

    // The value used to update the field specified by the lvalue.
    oneof rvalue {
      FieldValue field_rvalue = 2;
      HexstrValue hexstr_rvalue = 3;
      LookaheadValue lookahead_rvalue = 4;
      RExpression expression_rvalue = 5;
    }
  }

  oneof operation {
    Extract extract = 1;
    Set set = 3;
  }
}

// A single field in a transition key used in a parse state.
message ParserTransitionKeyField {
  oneof key_field {
    // Specifies a header field as part of the transition key.
    FieldValue field = 1;
    // Specifies the lookahead value as part of the transition key.
    // It evalutes to a set of bits from the input packet without advancing the
    // current read position.
    LookaheadValue lookahead = 2;
  }
}

// Defines a transition between two parse states.
message ParserTransition {
  oneof transition {
    // A default transition, applied when no other transitions are matched.
    DefaultTransition default_transition = 1;
    // A hex string transition, applied when the given hex string matches the
    // key.
    HexStringTransition hex_string_transition = 2;
  }
}

// Defines a default transition.
message DefaultTransition {
  // The name of the next state
  string next_state = 1;
}

// Defines a hex string transition.
message HexStringTransition {
  // Hex string value to be matched against the key.
  HexstrValue value = 1;
  // Hex string mask, to be ANDed with both the key and the value if not empty.
  HexstrValue mask = 2;
  // The name of the next state.
  string next_state = 3;
}

// Overall action structure, combining definition with implementation.
message Action {
  pdpi.IrActionDefinition action_definition = 1;
  ActionImplementation action_implementation = 2;
}

// Action implementation details.
message ActionImplementation {
  // Type of all action parameters.
  // Maps a parameter name to its bitwidth.
  map<string, int32> parameter_to_bitwidth = 2;
  // List of statements in action body in sequence.
  repeated Statement action_body = 3;
}

// Control flow information used during optimized symbolic execution
// (go/optimized-symbolic-execution) for each control point.
message OptimizedSymbolicExecutionInfo {
  // The name of the first control point where all execution paths branched
  // from the current control point will converge (refer to the respective field
  // in `CfgNode` for precise definition).
  string merge_point = 1;
  // Whether symbolic execution must continue (to the merge point) after
  // traversing the subgraph of the current control point (refer to the
  // respective field in `CfgNode` for details).
  bool continue_to_merge_point = 2;
}

// Overall table structure, combining definition with implementation.
message Table {
  // This contains the name, id, action list, field matches, and max size.
  pdpi.IrTableDefinition table_definition = 1;
  // This contains table implementation details.
  TableImplementation table_implementation = 2;
}

// Table implementation details that are not exposed in the definition.
message TableImplementation {
  enum ActionSelectorType {
    SIMPLE = 0;       // A Simple key and action table.
    INDIRECT = 1;     // Action profiles table.
    INDIRECT_WS = 2;  // Action profiles with dynamic selector.
  }

  // Table type with respect to action selection.
  ActionSelectorType action_selector_type = 1;

  // Default action name. This action is called with the default parameters
  // below, when none of this table's entries are matched on.
  string default_action = 2;

  // Hexstring values passed to default action as parameters.
  repeated string default_action_parameters = 3;

  // Maps every action used by this table (by name) to the next control
  // construct to be taken (also by name) if that action was matched.
  // All action names must be present in the map. In case the next control is
  // determined by whether the table is hit or not rather than by the executed
  // action, the map must contain exactly two special actions: "__HIT__" and
  // "__MISS__".
  map<string, string> action_to_next_control = 4;

  // Maps the name of a match (identical to the name used in p4 info), to
  // that match's target field.
  map<string, FieldValue> match_name_to_field = 5;

  // Control flow information for optimized symbolic execution.
  OptimizedSymbolicExecutionInfo optimized_symbolic_execution_info = 6;
}

// A conditional statement.
// In p4 and BMV2 JSON format, a condition is not part of an action body.
// It is a part of the top level control block. A conditional may be the first
// entry point in a workflow, or it may be called from other places using its
// name.
// Note that bmv2 accepts using a conditional name in fields that typically
// expect a table name, such as the initial_control, or next_controls
// attributes.
message Conditional {
  // The name is unique among conditionals AND tables.
  string name = 1;
  RValue condition = 2;
  // The next control points to execute for if and else branches (always
  // present).
  string if_branch = 3;
  string else_branch = 4;
  // Control flow information for optimized symbolic execution.
  OptimizedSymbolicExecutionInfo optimized_symbolic_execution_info = 5;
}

message Pipeline {
  string name = 1;
  // The name of the initial control construct.
  string initial_control = 2;
}

// Defines an error code that may be used in the program.
message Error {
  // The name of the error.
  string name = 1;
  // The value of the error code.
  uint32 value = 2;
}

// An abstract p4 statement corresponding to a top level operation within
// an action body.
message Statement {
  p4_symbolic.bmv2.SourceLocation source_info = 1;
  // Various statement concrete types.
  oneof statement {
    AssignmentStatement assignment = 2;
    DropStatement drop = 3;
    CloneStatement clone = 4;
    HashStatement hash = 5;
    ExitStatement exit = 6;
  }
}

// An assignment statement of the form <left> = <right>.
// Note that left can be an rvalue expression.
message AssignmentStatement {
  LValue left = 1;
  RValue right = 2;
}

// A statement marking the packet (by its header or metadata)
// to be dropped.
message DropStatement {
  HeaderValue header = 1;
}

// A statement cloning the packet, corresponds to `clone3`.
message CloneStatement {
  // We do not care about the arguments to clone3:
  // 1. We only support one version of cloning: ingree to egress
  // 2. The session id is not know at parsing time.
  // 3. We do not support clone calls preserving custom fields.
}

// Hashing statement to modify a selected field based on a hash calculation.
// Corresponds to calls to `hash`.
message HashStatement {
  // The field where the hash is stored.
  FieldValue field = 1;
  // We do not care about the remaining arguments (Hash algorithm,
  // range size, base value, and input fields).
  // These should perhaps be added in the future if the hash symbolic modeling
  // becomes more sophisticated.
}

// Statement that terminates the execution of all blocks currently executing.
message ExitStatement {}

// The structure of an abstract RValue.
message RValue {
  oneof rvalue {
    HeaderValue header_value = 1;
    FieldValue field_value = 2;
    HexstrValue hexstr_value = 3;
    BoolValue bool_value = 4;
    StringValue string_value = 5;
    Variable variable_value = 6;
    RExpression expression_value = 7;
  }
}

// The structure of an abstract LValue.
message LValue {
  oneof lvalue {
    FieldValue field_value = 1;
    Variable variable_value = 2;
  }
}

// A header accessed by name.
// This is used in operations and statements that refer to the whole packet.
// For example, in mark_to_drop(<header>).
message HeaderValue {
  string header_name = 1;
}

// A header field accessed statically.
// Field access that is dynamic (either in the index/offset or in the header
// object) is defined under AccessExpression below.
message FieldValue {
  string header_name = 1;
  string field_name = 2;
}

// Corresponds to a hex bigendiean encoding of a bit-string.
// In the bmv2 JSON format. The hexstring can be prefixed by "-"
// for negative values.
// We remove any such prefix and explicitly set the negative flag.
message HexstrValue {
  string value = 1;
  bool negative = 2;
}

// A boolean literal value.
message BoolValue {
  bool value = 1;
}

// A string literal value.
message StringValue {
  string value = 1;
}

// A lookahead value.
message LookaheadValue {
  // Bit offset of the lookahead operation.
  int32 offset = 1;
  // Bit width of the lookahead operation.
  int32 bitwidth = 2;
}

// A variable used within an LValue or RValue.
message Variable {
  string name = 1;
}

// An abstract RExpression.
message RExpression {
  oneof expression {
    BinaryExpression binary_expression = 1;
    UnaryExpression unary_expression = 2;
    TernaryExpression ternary_expression = 3;
    BuiltinExpression builtin_expression = 4;
  }
}

// Arithmetic, relational, and boolean binary operations.
message BinaryExpression {
  // Supported basic binary expressions.
  enum Operation {
    PLUS = 0;
    MINUS = 1;
    TIMES = 3;
    LEFT_SHIT = 4;    // <<
    RIGHT_SHIFT = 5;  // >>
    EQUALS = 6;       // ==
    NOT_EQUALS = 7;
    GREATER = 8;
    GREATER_EQUALS = 9;
    LESS = 10;
    LESS_EQUALS = 11;
    AND = 12;
    OR = 13;
    BIT_AND = 14;  // &
    BIT_OR = 15;
    BIT_XOR = 16;
  }

  Operation operation = 1;
  RValue left = 2;
  RValue right = 3;
}

// Arithmetic and boolean unary operations.
message UnaryExpression {
  // Supported basic unary expressions.
  enum Operation {
    NOT = 0;           // boolean not.
    BIT_NEGATION = 1;  // ~ (bitwise negation).
  }

  Operation operation = 1;
  RValue operand = 2;
}

// Ternary condition on the form <condition> ? <left> : <right>.
message TernaryExpression {
  RValue condition = 1;
  RValue left = 2;
  RValue right = 3;
}

// Calls tp built in functions (e.g. b2d).
message BuiltinExpression {
  // Supported built in functions.
  enum Function {
    BOOL_TO_DATA = 0;
    DATA_TO_BOOL = 1;
  }
  Function function = 1;
  repeated RValue arguments = 2;
}
